src/exponential_utils.jl

```diff
@@ -0,0 +1,216 @@
+  v::AbstractVector{T}, k::Integer; cache=nothing) where {T <: Number}
+  @assert size(w, 1) == size(A, 1) == size(A, 2) == length(v) "Dimension mismatch"
+  @assert size(w, 2) == k+1 "Dimension mismatch"
+  m = length(v)
+  # Construct the extended matrix
+  if cache == nothing
+    cache = zeros(T, m+k, m+k)
+  else
+    @assert size(cache) == (m+k, m+k) "Dimension mismatch"
+    fill!(cache, zero(T))
+  end
+  cache[1:m, 1:m] = A
+  cache[1:m, m+1] = v
+  for i = m+1:m+k-1
+    cache[i, i+1] = one(T)
+  end
+  P = Base.LinAlg.expm!(cache)
+  # Extract results
+  @views A_mul_B!(w[:, 1], P[1:m, 1:m], v)
+  for i = 1:k
+    w[:, i+1] = P[1:m, m+i]
+  end
+  return w
+end
+
+"""
+    phim(A,k[;cache]) -> [phi_0(A),phi_1(A),...,phi_k(A)]
+
+Compute the matrix phi functions for all orders up to k.
+
+The phi functions are defined as
+  
+```math
+\\varphi_0(z) = \\exp(z),\\quad \\varphi_k(z+1) = \\frac{\\varphi_k(z) - 1}{z} 
+```
+
+Calls `phimv_dense` on each of the basis vectors to obtain the answer.
+"""
+function phim(A::AbstractMatrix{T}, k::Integer; cache=nothing) where {T <: Number}
+  m = size(A, 1)
+  out = [Matrix{T}(m, m) for i = 1:k+1]
+  E = eye(A)
+  w = Matrix{T}(m, k+1)
+  if cache == nothing
+    cache = Matrix{T}(m+k, m+k)
+  end
+  for i = 1:m
+    e = E[:, i]
+    phimv_dense!(w, A, e, k; cache=cache) # w = [phi_0(A)*e phi_1(A)*e ... phi_k(A)*e]
+    for j = 1:k+1
+      out[j][:, i] = w[:, j]
+    end
+  end
+  return out
+end
+
+##############################################
+# Krylov algorithms
+
+"""
+    arnoldi(A,b,m) -> V,H
+
+Performs `m` anoldi iterations to obtain the Krylov subspace K_m(A,b).
+
+The output is the n x m unitary basis vectors `V` and the m x m upper 
+Heisenberg matrix `H`. They are related by the recurrence formula
+
+```
+v_1=b,\\quad Av_j = \\sum_{i=1}^{j+1}h_{ij}v_i\\quad(j = 1,2,\\ldots,m)
+```
+"""
+function arnoldi(A, b, m; cache=nothing)
+  V = Matrix{eltype(b)}(length(b), m)
+  H = Matrix{eltype(b)}(m, m)
+  arnoldi!(V, H, A, b, m; cache=cache)
+end
+"""
+    arnoldi!(V,H,A,b,m) -> V,H
+
+Non-allocating version of `arnoldi`.
+"""
+function arnoldi!(V::Matrix{T}, H::Matrix{T}, A, b::AbstractVector{T}, 
+  m::Integer; cache=nothing) where {T <: Number}
+  @assert size(V,1) == size(A,1) == size(A,2) == length(b) "Dimension mismatch"
+  @assert size(V,2) == size(H,1) == size(H,2) == m "Dimension mismatch"
+  if cache == nothing
+    cache = similar(b)
+  else
+    @assert size(cache) == (length(b),)
+  end
+  V[:, 1] = normalize(b)
+  for j = 1:m-1
+    A_mul_B!(cache, A, V[:, j])
+    for i = 1:j
+      alpha = dot(V[:, i], cache)
+      H[i, j] = alpha
+      Base.axpy!(-alpha, V[:, i], cache)
+    end
+    beta = norm(cache)
+    H[j+1, j] = beta
+    V[:, j+1] = cache / beta
+  end
+  # Last iteration (j = m)
+  A_mul_B!(cache, A, V[:, m-1])
+  for i = 1:m
+    alpha = dot(V[:, i], cache)
+    H[i, m] = alpha
+    Base.axpy!(-alpha, V[:, i], cache) # is this still needed?

```

Comment:
Why the comment `# is this still needed?`?

Reply:
I changed the implementation of `arnoldi` a bit to not return the extended matrices. Which means `v_m+1` is not needed. And since the previous `v`s are in theory orthogonal this step is not needed if the dot products are all that is needed.
Of course arnoldi does not yield perfectly orthogonal vectors, so it makes sense to play it safe.